presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -351,7 +352,7 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
 
                 BucketNodeMap bucketNodeMap;
                 List<Node> stageNodeList;
-                if (plan.getSubStages().isEmpty()) {
+                if (plan.getSubStages().isEmpty() || plan.getFragment().getRemoteSourceNodes().stream().allMatch(node -> node.getExchangeType() == REPLICATE)) {
```

Comment:
What about 

```java
                if (plan.getFragment().getRemoteSourceNodes().stream().noneMatch(node -> node.getExchangeType() != REPLICATE))
``
?

ditto for the code in `PlanFragmenter.java`

Reply:
thanks! didn't realize that an empty stream would return true, but it makes sense :).